providers/dns/arvancloud/internal/client.go

```diff
@@ -0,0 +1,200 @@
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() { _ = resp.Body.Close() }()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response body: %w", err)
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("could not get records %s: Domain: %s; Status: %s; Body: %s",
+			search, domain, resp.Status, string(body))
+	}
+
+	response := &apiResponse{}
+	err = json.Unmarshal(body, response)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode response body: %w", err)
+	}
+
+	var records []DNSRecord
+	err = json.Unmarshal(response.Data, &records)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode records: %w", err)
+	}
+
+	return records, nil
+}
+
+// CreateRecord creates a DNS record.
+// https://www.arvancloud.com/docs/api/cdn/4.0#operation/dns_records.create
+func (c *Client) CreateRecord(domain string, record DNSRecord) (*DNSRecord, error) {
+	reqBody, err := json.Marshal(record)
+	if err != nil {
+		return nil, err
+	}
+
+	endpoint, err := c.createEndpoint("cdn", "4.0", "domains", domain, "dns-records")
+	if err != nil {
+		return nil, fmt.Errorf("failed to create endpoint: %w", err)
+	}
+
+	resp, err := c.do(http.MethodPost, endpoint.String(), bytes.NewReader(reqBody))
+	if err != nil {
+		return nil, err
+	}
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read response body: %w", err)
+	}
+
+	if resp.StatusCode != http.StatusCreated {
+		return nil, fmt.Errorf("could not create record %s; Domain: %s; Status: %s; Body: %s", string(reqBody), domain, resp.Status, string(body))
+	}
+
+	response := &apiResponse{}
+	err = json.Unmarshal(body, response)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode response body: %w", err)
+	}
+
+	var newRecord DNSRecord
+	err = json.Unmarshal(response.Data, &newRecord)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode record: %w", err)
+	}
+
+	return &newRecord, nil
+}
+
+// DeleteRecord deletes a DNS record.
+// https://www.arvancloud.com/docs/api/cdn/4.0#operation/dns_records.remove
+func (c *Client) DeleteRecord(domain, id string) error {
+	endpoint, err := c.createEndpoint("cdn", "4.0", "domains", domain, "dns-records", id)
+	if err != nil {
+		return fmt.Errorf("failed to create endpoint: %w", err)
+	}
+
+	resp, err := c.do(http.MethodDelete, endpoint.String(), nil)
+	if err != nil {
+		return err
+	}
+
+	if resp.StatusCode != http.StatusOK {
+		body, _ := ioutil.ReadAll(resp.Body)
+		return fmt.Errorf("could not delete record %s; Domain: %s; Status: %s; Body: %s", id, domain, resp.Status, string(body))
+	}
+
+	return nil
+}
+
+func (c *Client) do(method, endpoint string, body io.Reader) (*http.Response, error) {
+	req, err := http.NewRequest(method, endpoint, body)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set("Accept", "application/json")
+	req.Header.Set("Content-Type", "application/json")
```

Comment:
this is cause the 500 error, `GET` should not have this header.

```
	if body != nil {
		req.Header.Set("Content-Type", "application/json")
	}
```

Reply:
There is no GET so the error is not related to that.

Reply:
I improved the error log, could you retry?

Reply:
I may be wrong, but what I saw was that if we want to delete a record and the given id is not valid, arvancloud return 500 page .

Reply:
here:

```
2020/05/28 18:54:18 No key found for account hi@hashemian.me. Generating a P384 key.
2020/05/28 18:54:18 Saved key to /home/jimc/go/src/github.com/go-acme/lego/cmd/lego/.lego/accounts/acme-staging-v02.api.letsencrypt.org/hi@hashemian.me/keys/hi@hashemian.me.key
2020/05/28 18:54:21 [INFO] acme: Registering account for hi@hashemian.me
!!!! HEADS UP !!!!

Your account credentials have been saved in your Let's Encrypt
configuration directory at "/home/jimc/go/src/github.com/go-acme/lego/cmd/lego/.lego/accounts".

You should make a secure backup of this folder now. This
configuration directory will also contain certificates and
private keys obtained from Let's Encrypt so making regular
backups of this folder is ideal.
2020/05/28 18:54:22 [INFO] [*.thetangle.ir, thetangle.ir] acme: Obtaining bundled SAN certificate
2020/05/28 18:54:25 [INFO] [*.thetangle.ir] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/59941157
2020/05/28 18:54:25 [INFO] [thetangle.ir] AuthURL: https://acme-staging-v02.api.letsencrypt.org/acme/authz-v3/59941158
2020/05/28 18:54:25 [INFO] [*.thetangle.ir] acme: use dns-01 solver
2020/05/28 18:54:25 [INFO] [thetangle.ir] acme: Could not find solver for: tls-alpn-01
2020/05/28 18:54:25 [INFO] [thetangle.ir] acme: Could not find solver for: http-01
2020/05/28 18:54:25 [INFO] [thetangle.ir] acme: use dns-01 solver
2020/05/28 18:54:25 [INFO] [*.thetangle.ir] acme: Preparing to solve DNS-01
2020/05/28 18:54:38 [INFO] [thetangle.ir] acme: Preparing to solve DNS-01
2020/05/28 18:54:44 [INFO] [*.thetangle.ir] acme: Trying to solve DNS-01
2020/05/28 18:54:44 [INFO] [*.thetangle.ir] acme: Checking DNS record propagation using [8.8.8.8:53 8.8.4.4:53 192.168.43.202:53]
2020/05/28 18:54:44 [INFO] Wait for propagation [timeout: 2m0s, interval: 2s]
2020/05/28 18:54:46 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 18:54:50 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 18:54:54 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 18:54:58 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 18:55:03 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 18:55:07 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 18:55:11 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 18:55:14 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 18:55:17 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 18:55:24 [INFO] [*.thetangle.ir] acme: Waiting for DNS record propagation.
2020/05/28 18:55:29 [INFO] [*.thetangle.ir] The server validated our request
2020/05/28 18:55:29 [INFO] [thetangle.ir] acme: Trying to solve DNS-01
2020/05/28 18:55:29 [INFO] [thetangle.ir] acme: Checking DNS record propagation using [8.8.8.8:53 8.8.4.4:53 192.168.43.202:53]
2020/05/28 18:55:29 [INFO] Wait for propagation [timeout: 2m0s, interval: 2s]
2020/05/28 18:55:32 [INFO] [thetangle.ir] The server validated our request
2020/05/28 18:55:32 [INFO] [*.thetangle.ir] acme: Cleaning DNS-01 challenge
2020/05/28 18:55:33 [WARN] [*.thetangle.ir] acme: cleaning up failed: arvancloud: failed to delate TXT record: id=5ace4bfb-014c-4362-8f40-0e4e50569118: could not delete record 5ace4bfb-014c-4362-8f40-0e4e50569118; Domain: thetangle.ir; Status: 500 Internal Server Error; Body: {
    "status": false,
    "message": "Internal Server Error"
} 
2020/05/28 18:55:33 [INFO] [thetangle.ir] acme: Cleaning DNS-01 challenge
2020/05/28 18:55:34 [WARN] [thetangle.ir] acme: cleaning up failed: arvancloud: failed to delate TXT record: id=27e49a84-1df4-4bac-ac96-8d7f33cfcb64: could not delete record 27e49a84-1df4-4bac-ac96-8d7f33cfcb64; Domain: thetangle.ir; Status: 500 Internal Server Error; Body: {
    "status": false,
    "message": "Internal Server Error"
} 
2020/05/28 18:55:34 [INFO] [*.thetangle.ir, thetangle.ir] acme: Validations succeeded; requesting certificates
2020/05/28 18:55:36 [INFO] [*.thetangle.ir] Server responded with a certificate.
```

I'm pretty sure my suggestion fixes it

Reply:
There are no GET request so it's impossible.

But maybe you are speaking about DELETE, but I have not changed this part in my review/merge.

Reply:
> I may be wrong, but what I saw was that if we want to delete a record and the given id is not valid, arvancloud return 500 page .

I think that @FaRzad-845 is right.

Reply:
i add this line in my codes 

```if method != http.MethodDelete {
		req.Header.Set("Content-Type", "application/json")
	}```

Reply:
yes, I meant `DELETE`, that was my mistake, I've fixed it later (which accidentally removed your changes too)

Reply:
I fixed that, could you retry once again?